theLudovyc/TheFactory.gd

```diff
@@ -1,160 +1,103 @@
-		)
-
-		storage.update_global_production_rate(resource_type)
 
 
 func population_increase(amount: int):
-	var population_pool = amount
+	var population_pool: int = amount
 
 	var i := 0
 
-	while i < waiting_lines.size():
-		if waiting_lines[i][Waiting_Lines.needed_workers] <= population_pool:
-			var waiting_line = waiting_lines.pop_at(i)
-
-			var waiting_line_needed_workers = waiting_line[Waiting_Lines.needed_workers]
-
-			create_or_update_line(
-				waiting_line[Waiting_Lines.resource_type], waiting_line_needed_workers
-			)
-
-			population_pool -= waiting_line_needed_workers
-
+	while i < waiting_line_buildings.size():
+		var waiting_building: BuildingModel = waiting_line_buildings[i]
+		var needed_worker: int = waiting_building.needed_worker()
+		if needed_worker <= population_pool:
+			waiting_line_buildings.remove_at(i)
+			population_pool = population_pool - needed_worker
+			add_workers_to_model(waiting_building)
 			if population_pool <= 0:
 				break
 		else:
-			i += 1
+			i = i + 1
 
 
 func population_decrease(amount: int):
-	var population_pool = amount
-
-	while not production_lines.is_empty():
-		var resource_type = production_lines.keys().pick_random()
-
-		var line = production_lines[resource_type]
-
-		var current_workers = line[Production_Line.current_workers]
-
-		if current_workers >= amount:
-			line[Production_Line.current_workers] -= amount
-
-			waiting_lines.push_back([resource_type, amount])
-
-			population_pool -= amount
-		else:
-			line[Production_Line.current_workers] = 0
-
-			waiting_lines.push_back([resource_type, current_workers])
-
-			population_pool -= current_workers
-
-		if line[Production_Line.current_workers] == 0:
-			production_lines.erase(resource_type)
-
-		storage.update_global_production_rate(resource_type)
-
+	var population_pool: int = amount
+	
+	while not buildings.is_empty():
+		var building: BuildingModel = buildings.pick_random()
+		var current_workes: int = building.current_worker
+		if current_workes > 0:
+			waiting_line_buildings.push_back(building)
+			population_pool = population_pool - amount
+			building.rem_all_worker()
 		if population_pool <= 0:
 			break
 
 
 func _on_TheTicker_tick():
-	for resource_type in production_lines:
-		var line = production_lines[resource_type]
-
-		line[Production_Line.current_ticks] += 1
-
-		if line[Production_Line.current_ticks] >= Recipes.get_recipe_needed_ticks(resource_type):
-			line[Production_Line.current_ticks] = 0
-
-			storage.add_resource(resource_type, line[Production_Line.production_rate])
+	for single_building in buildings:
+		if is_instance_of(single_building, ProductionBuildingModel):
+			var production_building: ProductionBuildingModel = single_building as ProductionBuildingModel
+			production_building.singleTick(storage)
+			
+			
+func find_building_by_id(id: String) -> BuildingModel:
+	for building_model in buildings:
+		if building_model.id == id:
+			return building_model
```

Comment:
This algorithm is not efficient. Maybe you can make a dictionary [id]=building_model .